async/compiler/src/openxla/compiler/Dialect/Async/Conversion/ConvertAsyncToRuntime.cpp

```diff
@@ -0,0 +1,253 @@
+// AsyncAPI for importing Async VM module functions
+//===----------------------------------------------------------------------===//
+
+class AsyncAPI {
+ public:
+  // Imports `@async.token.await` into the module
+  func::FuncOp getTokenAwait(PatternRewriter &rewriter, ModuleOp module);
+  // Imports `@async.value.await.i32` into the module
+  func::FuncOp getValueAwaitI32(PatternRewriter &rewriter, ModuleOp module);
+  // Imports `@async.value.await.ref` into the module
+  func::FuncOp getValueAwaitRef(PatternRewriter &rewriter, ModuleOp module);
+
+  SymbolTable &symTable(ModuleOp module);
+
+  bool isScalarType(Type type) {
+    return type.isInteger(32) || type.isInteger(64) || type.isF32() ||
+           type.isF64();
+  }
+
+ private:
+  func::FuncOp addDecl(PatternRewriter &rewriter, ModuleOp module,
+                       StringAttr name, FunctionType function_type);
+  SymbolTableCollection symTable_;
+};
+
+SymbolTable &AsyncAPI::symTable(ModuleOp module) {
+  return symTable_.getSymbolTable(module);
+}
+
+func::FuncOp AsyncAPI::addDecl(PatternRewriter &rewriter, ModuleOp module,
+                               StringAttr name, FunctionType function_type) {
+  if (auto fn = symTable_.lookupNearestSymbolFrom<func::FuncOp>(module, name))
+    return fn;
+
+  ImplicitLocOpBuilder b(UnknownLoc::get(module->getContext()), rewriter);
+  b.setInsertionPointToEnd(module.getBody());
+
+  auto fn = b.create<func::FuncOp>(name, function_type);
+  fn.setPrivate();
+  symTable(module).insert(fn);
+  return fn;
+}
+
+func::FuncOp AsyncAPI::getTokenAwait(PatternRewriter &rewriter,
+                                     ModuleOp module) {
+  MLIRContext *ctx = module->getContext();
+  SmallVector<Type> args{ValueType::get(ctx)};
+  auto functionType = FunctionType::get(ctx, args, /*rets=*/{});
+
+  return addDecl(rewriter, module,
+                 StringAttr::get(ctx, "async.value.await.token"), functionType);
+}
+
+func::FuncOp AsyncAPI::getValueAwaitI32(PatternRewriter &rewriter,
+                                        ModuleOp module) {
+  MLIRContext *ctx = module->getContext();
+  SmallVector<Type> args{ValueType::get(ctx)};
+  SmallVector<Type> rets{IntegerType::get(ctx, 32)};
+  auto functionType = FunctionType::get(ctx, args, rets);
+
+  return addDecl(rewriter, module,
+                 StringAttr::get(ctx, "async.value.await.i32"), functionType);
+}
+
+func::FuncOp AsyncAPI::getValueAwaitRef(PatternRewriter &rewriter,
+                                        ModuleOp module) {
+  MLIRContext *ctx = module->getContext();
+  SmallVector<Type> args{ValueType::get(ctx)};
+  SmallVector<Type> rets{IREE::Util::ObjectType::get(ctx)};
+  auto functionType = FunctionType::get(ctx, args, rets);
+
+  return addDecl(rewriter, module,
+                 StringAttr::get(ctx, "async.value.await.ref"), functionType);
+}
+
+class FuncOpConversion : public OpConversionPattern<func::FuncOp> {
```

Comment:
I believe that https://mlir.llvm.org/doxygen/namespacemlir.html#a12287a4adc9d89ce2ed49712b091c5fa does this already, there are many examples in MLIR passes

Reply:
Fixed